sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala

```diff
@@ -0,0 +1,200 @@
+    if (SQLConf.get.nestedSchemaPruningEnabled) {
+      apply0(plan)
+    } else {
+      plan
+    }
+
+  private def apply0(plan: LogicalPlan): LogicalPlan =
+    plan transformDown {
+      case op @ PhysicalOperation(projects, filters,
+          l @ LogicalRelation(hadoopFsRelation @ HadoopFsRelation(_, _,
+            dataSchema, _, _: ParquetFileFormat, _), _, _, _)) =>
+        val projectionRootFields = projects.flatMap(getRootFields)
+        val filterRootFields = filters.flatMap(getRootFields)
+        val requestedRootFields = (projectionRootFields ++ filterRootFields).distinct
+
+        // If requestedRootFields includes a nested field, continue. Otherwise,
+        // return op
+        if (requestedRootFields.exists { case RootField(_, derivedFromAtt) => !derivedFromAtt }) {
+          val prunedDataSchema = buildPrunedDataSchema(dataSchema, requestedRootFields)
+
+          // If the data schema is different from the pruned data schema, continue. Otherwise,
+          // return op. We effect this comparison by counting the number of "leaf" fields in
+          // each schemata, assuming the fields in [[prunedDataSchema]] are a subset of the fields
+          // in dataSchema.
+          if (countLeaves(dataSchema) > countLeaves(prunedDataSchema)) {
+            val prunedParquetRelation =
+              hadoopFsRelation.copy(dataSchema = prunedDataSchema)(hadoopFsRelation.sparkSession)
+
+            val prunedRelation = buildPrunedRelation(l, prunedParquetRelation)
+            val projectionOverSchema = ProjectionOverSchema(prunedDataSchema)
+
+            // Construct a new target for our projection by rewriting and
+            // including the original filters where available
+            val projectionChild =
+              if (filters.nonEmpty) {
+                val projectedFilters = filters.map(_.transformDown {
+                  case projectionOverSchema(expr) => expr
+                })
+                val newFilterCondition = projectedFilters.reduce(And)
+                Filter(newFilterCondition, prunedRelation)
+              } else {
+                prunedRelation
+              }
+
+            // Construct the new projections of our Project by
+            // rewriting the original projections
+            val newProjects = projects.map(_.transformDown {
+              case projectionOverSchema(expr) => expr
+            }).map { case expr: NamedExpression => expr }
+
+            if (log.isDebugEnabled) {
+              logDebug(s"New projects:\n${newProjects.map(_.treeString).mkString("\n")}")
+              logDebug(s"Pruned data schema:\n${prunedDataSchema.treeString}")
+            }
+
+            Project(newProjects, projectionChild)
+          } else {
+            op
+          }
+        } else {
+          op
+        }
+    }
+
+  private def buildPrunedDataSchema(fileDataSchema: StructType,
+                                    requestedRootFields: Seq[RootField]) = {
```

Comment:
Will fix all this. Seems like this is an opportunity to catch some of this with static analysis to avoid wasting the reviewer's time. Is Scalastyle not configured to catch this stuff?

Reply:
As far as I know, no. I use IntelliJ (integration with scalastyle) to catch basic style nits but rather then that, I do it manually. It took me a while to get used to it before FWIW.